File: src/main/java/com/springdebugger/llm/LlmDiagnosisEngine.java
package com.springdebugger.llm;

import com.springdebugger.engine.DiagnosisEngine;
import com.springdebugger.model.Confidence;
import com.springdebugger.model.DiagnosisCard;
import com.springdebugger.model.RawSignal;
import org.yaml.snakeyaml.Yaml;

import java.util.Map;
import java.util.Optional;

/**
 * LLM fallback engine (M13). Used only when the rule engine returns no match (the pipeline's
 * last resort). Builds a tightly scoped prompt from the extracted signal, asks a local Ollama
 * model for a one-sentence diagnosis and one-sentence fix as JSON, and parses the reply into a
 * {@link DiagnosisCard} labelled {@code ruleId="llm"} at MEDIUM confidence.
 *
 * <p>Safety contract (PLAN 8.3): if the response cannot be parsed into both fields, nothing is
 * shown. A malformed model reply must never surface as a confident diagnosis. LLM output is
 * always MEDIUM, never HIGH, so it ranks below any rule match and is clearly distinguishable.
 */
public final class LlmDiagnosisEngine implements DiagnosisEngine {

    private final OllamaClient client;

    public LlmDiagnosisEngine(OllamaClient client) {
        this.client = client;
    }

    @Override
    public Optional<DiagnosisCard> diagnose(RawSignal signal) {
        if (signal == null) return Optional.empty();
        String prompt = buildPrompt(signal);
        Optional<String> reply = client.generate(prompt);
        return reply.flatMap(r -> parseCard(r, signal));
    }

    static String buildPrompt(RawSignal signal) {
        StringBuilder sb = new StringBuilder();
        sb.append("You are a Spring Boot error triage assistant. ")
                .append("Given the extracted signal from a failing Spring Boot application, ")
                .append("respond ONLY with a JSON object of exactly two string fields: ")
                .append("\"diagnosis\" (one sentence naming the precise problem) and ")
                .append("\"fix\" (one sentence giving the single corrective action). ")
                .append("Do not include any other text.\n\n");
        sb.append("Phase: ").append(signal.getPhase()).append('\n');
        if (signal.getDeepestCausedByClass() != null) {
            sb.append("Deepest exception: ").append(signal.getDeepestCausedByClass()).append('\n');
        }
        if (signal.getDeepestCausedByMessage() != null) {
            sb.append("Exception message: ").append(signal.getDeepestCausedByMessage()).append('\n');
        }
        if (signal.getBannerDescription() != null) {
            sb.append("Failure analyzer description: ").append(signal.getBannerDescription()).append('\n');
        }
        if (signal.getBannerAction() != null) {
            sb.append("Failure analyzer action: ").append(signal.getBannerAction()).append('\n');
        }
        if (signal.getFailingBeanName() != null) {
            sb.append("Failing bean: ").append(signal.getFailingBeanName()).append('\n');
        }
        sb.append("\nLog excerpt:\n").append(truncate(signal.getRawExcerpt(), 4000));
        return sb.toString();
    }

    /** Parses the model's JSON reply. Returns empty unless BOTH diagnosis and fix are present. */
    static Optional<DiagnosisCard> parseCard(String reply, RawSignal signal) {
        if (reply == null || reply.isBlank()) return Optional.empty();
        try {
            Object parsed = new Yaml().load(reply);
            if (!(parsed instanceof Map<?, ?> map)) return Optional.empty();
            Object diagnosis = map.get("diagnosis");
            Object fix = map.get("fix");
            if (!(diagnosis instanceof String d) || !(fix instanceof String f)) return Optional.empty();
            if (d.isBlank() || f.isBlank()) return Optional.empty();
            return Optional.of(new DiagnosisCard(
                    "llm",
                    signal.getPhase(),
                    d.trim(),
                    f.trim(),
                    Confidence.MEDIUM,
                    signal.getRawExcerpt()));
        } catch (RuntimeException e) {
            return Optional.empty();
        }
    }

    private static String truncate(String s, int max) {
        if (s == null) return "";
        return s.length() <= max ? s : s.substring(0, max) + "\n...[truncated]";
    }
}

File: vscode-extension/src/engine/llm-diagnosis-engine.ts
// Port of com.springdebugger.llm.LlmDiagnosisEngine. Used only when the rule engine returns no
// match. Builds a tightly scoped prompt from the signal, asks a local Ollama model for a one-sentence
// diagnosis and fix as JSON, and parses the reply into a card labelled ruleId="llm" at MEDIUM
// confidence. Safety contract: if the reply cannot be parsed into BOTH fields, nothing is shown.
import { DiagnosisEngine } from './enricher';
import { DiagnosisCard, RawSignal } from './models';
import { OllamaClient } from './ollama-http-client';

export class LlmDiagnosisEngine implements DiagnosisEngine {
  constructor(private readonly client: OllamaClient) {}

  async diagnose(signal: RawSignal): Promise<DiagnosisCard | null> {
    if (signal === null || signal === undefined) return null;
    const reply = await this.client.generate(buildPrompt(signal));
    return reply === null ? null : parseCard(reply, signal);
  }
}

export function buildPrompt(signal: RawSignal): string {
  let sb =
    'You are a Spring Boot error triage assistant. ' +
    'Given the extracted signal from a failing Spring Boot application, ' +
    'respond ONLY with a JSON object of exactly two string fields: ' +
    '"diagnosis" (one sentence naming the precise problem) and ' +
    '"fix" (one sentence giving the single corrective action). ' +
    'Do not include any other text.\n\n';
  sb += 'Phase: ' + signal.phase + '\n';
  if (signal.deepestCausedByClass !== null) sb += 'Deepest exception: ' + signal.deepestCausedByClass + '\n';
  if (signal.deepestCausedByMessage !== null) sb += 'Exception message: ' + signal.deepestCausedByMessage + '\n';
  if (signal.bannerDescription !== null) sb += 'Failure analyzer description: ' + signal.bannerDescription + '\n';
  if (signal.bannerAction !== null) sb += 'Failure analyzer action: ' + signal.bannerAction + '\n';
  if (signal.failingBeanName !== null) sb += 'Failing bean: ' + signal.failingBeanName + '\n';
  sb += '\nLog excerpt:\n' + truncate(signal.rawExcerpt, 4000);
  return sb;
}

/** Parses the model's JSON reply. Returns null unless BOTH diagnosis and fix are present. */
export function parseCard(reply: string | null, signal: RawSignal): DiagnosisCard | null {
  if (reply === null || reply.trim() === '') return null;
  try {
    const parsed = JSON.parse(reply);
    if (!parsed || typeof parsed !== 'object') return null;
    const diagnosis = (parsed as Record<string, unknown>).diagnosis;
    const fix = (parsed as Record<string, unknown>).fix;
    if (typeof diagnosis !== 'string' || typeof fix !== 'string') return null;
    if (diagnosis.trim() === '' || fix.trim() === '') return null;
    return {
      ruleId: 'llm',
      phase: signal.phase,
      diagnosisSentence: diagnosis.trim(),
      fixSentence: fix.trim(),
      confidence: 'MEDIUM',
      excerpt: signal.rawExcerpt,
    };
  } catch {
    return null;
  }
}

function truncate(s: string | null, max: number): string {
  if (s === null) return '';
  return s.length <= max ? s : s.substring(0, max) + '\n...[truncated]';
}
